phoenix-core/src/it/java/org/apache/phoenix/end2end/WALAnnotationIT.java

```diff
@@ -456,6 +456,42 @@ public void testTenantViewUpsertWithIndex() throws Exception {
         tenantViewHelper(true);
     }
 
+    @Test
+    public void testOnDuplicateUpsertWithIndex() throws Exception {
+        if (this.isImmutable) {
+            return; // on duplicate is not supported for immutable tables
```

Comment:
Do we error out when the table is immutable?
Also does conditional update work with single cell format? I recommend adding a test for it

Reply:
Yes, for immutable tables the Upsert Compiler throws an exception if ON DUPLICATE KEY is used.